linkerd/access-log/src/layer.rs

```diff
@@ -0,0 +1,122 @@
+use crate::AccessLog;
+use futures::{ready, TryFuture};
+use pin_project::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::SystemTime;
+use tokio::sync::mpsc;
+use tracing::warn;
+
+/// A layer that adds access logging
+#[derive(Clone)]
+pub struct AccessLogLayer {
+    sink: Option<mpsc::Sender<AccessLog>>,
+}
+
+#[derive(Clone)]
+pub struct AccessLogContext<Svc> {
+    inner: Svc,
+    sink: Option<mpsc::Sender<AccessLog>>,
+}
+
+#[pin_project]
+pub struct ResponseFuture<F> {
+    state: Option<(AccessLog, mpsc::Sender<AccessLog>)>,
+
+    #[pin]
+    inner: F,
+}
+
+impl<Svc> tower::layer::Layer<Svc> for AccessLogLayer {
+    type Service = AccessLogContext<Svc>;
+
+    fn layer(&self, inner: Svc) -> Self::Service {
+        Self::Service {
+            inner,
+            sink: self.sink.clone(),
+        }
+    }
+}
+
+impl<Svc, B1, B2> tower::Service<http::Request<B1>> for AccessLogContext<Svc>
+where
+    Svc: tower::Service<http::Request<B1>, Response = http::Response<B2>>,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = ResponseFuture<Svc::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Svc::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: http::Request<B1>) -> Self::Future {
+        let sink = match &self.sink {
+            Some(sink) => sink,
+            None => {
+                return ResponseFuture {
+                    state: None,
+                    inner: self.inner.call(request),
+                }
+            }
+        };
+
+        let t0 = SystemTime::now();
+
+        let host = request.headers().get("Host").cloned();
+
+        let trace_id = request
+            .headers()
+            .get("X-Amzn-Trace-Id")
+            .or_else(|| request.headers().get("X-Request-ID"))
```

Comment:
we should also handle B3 trace propagation, especially since it's the only trace propagation header that Linkerd proxies are [actually aware of](https://github.com/linkerd/linkerd2-proxy/blob/1b3e899efad2ad4ee471c5b49482437e7a325bde/linkerd/trace-context/src/propagation.rs)...

Reply:
Haha, yes definitely! In fact I'll just use that header, no need to unnecessarily pollute linkerd2 with Amazon specific headers. Longer term might be nice to have the ability to include arbitrary headers, maybe as part of a [custom log format system](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage), but definitely not something for an MVP implementation such as this.